### PR TITLE
Add staging cleanup script

### PR DESCRIPTION
**Pull request by AyodeAwe**
PR adds a workflow that runs each night and cleans up any rapidsai/staging images that are older than 30 days.

### CODE REVIEW

File: ci/cleanup_images.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+echo "$tags_json" | jq -c '.results[]' | while read -r tag_info; do
+    last_pushed=$(echo "$tag_info" | jq -r '.tag_last_pushed')
+    last_pushed_ts=$(date --date="$last_pushed" +%s)
+
+    #age in days
+    current_time=$(date +%s)
+    age=$(( (current_time - last_pushed_ts) / 86400 ))
+
+    if [ "$age" -gt 30 ]; then
```
Comment (ajschmidt8) [inline ci/cleanup_images.sh:39]:
```suggestion
    current_time=$(date +%s)
    age_in_days=$(( (current_time - last_pushed_ts) / 86400 )) # 86400 seconds in one day

    if [ "$age_in_days" -gt 30 ]; then
```

[Magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) like `86400` are ambiguous.

We can eliminate the ambiguity by adding a comment at the end of the line.

We can also eliminate `age in days` comment by just renaming the variable to `age_in_days`.

**Review by ajschmidt8 — COMMENTED** [review]
Left a few comments.

Do you know the DockerHub API imposes a rate-limit?

If so, we might want to add an arbitrary `sleep 1s` in the `while` loop to prevent any issues.

File: ci/cleanup_images.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+repo="staging"
+
+delete_image() {
+    local org=$1
+    local repo=$2
+    local tag=$3
+
+    curl --fail-with-body -i -X DELETE \
+        -H "Accept: application/json" \
+        -H "Authorization: JWT $HUB_TOKEN" \
```
[the remark this answers is not in the document]
Comment (AyodeAwe) [reply ci/cleanup_images.sh:13]: PR is here: https://github.com/rapidsai/shared-actions/pull/4

**Comment by AyodeAwe** [comment]
> Left a few comments.
> 
> Do you know the DockerHub API imposes a rate-limit?
> 
> If so, we might want to add an arbitrary `sleep 1s` in the `while` loop to prevent any issues.

I don't think they do for these API calls

**Comment by AyodeAwe** [comment]
> > Left a few comments.
> > Do you know the DockerHub API imposes a rate-limit?
> > If so, we might want to add an arbitrary `sleep 1s` in the `while` loop to prevent any issues.
> 
> I don't think they do for these API calls

I stand corrected. Looks like they do: 600 requests per minute from what I see in the response headers

File: ci/cleanup_images.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    curl -L --fail-with-body \
+        -H "Accept: application/json" \
+        -H "Authorization: JWT $HUB_TOKEN" \
+        "https://hub.docker.com/v2/namespaces/$org/repositories/$repo/tags?page=$page&page_size=100"
+}
+
+page=1
+next_page="non-null"
+
+while [ -n "$next_page" ] && [ "$next_page" != "null" ]; do
```
Comment (ajschmidt8) [inline ci/cleanup_images.sh:34]:
```suggestion
while [ "$next_page" != "null" ]; do
```

We shouldn't need this conditional statement.

File: ci/cleanup_images.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        last_pushed_ts=$(date --date="$last_pushed" +%s)
+
+        current_time=$(date +%s)
+        age_in_days=$(( (current_time - last_pushed_ts) / 86400 )) # 86400 seconds in one day
+
+        if [ "$age_in_days" -gt 30 ]; then
+            tag_name=$(echo "$tag_info" | jq -r '.name')
+            delete_image "$org" "$repo" "$tag_name"
+        fi
+    done
```
Comment (ajschmidt8) [inline ci/cleanup_images.sh:47]:
```suggestion
echo "$tags_json" | jq -c '.results[]' | while read -r tag_info; do
        tag_name=$(echo "$tag_info" | jq -r '.name')
        last_pushed=$(echo "$tag_info" | jq -r '.tag_last_pushed')
        last_pushed_ts=$(date --date="$last_pushed" +%s)

        current_time=$(date +%s)
        age_in_days=$(( (current_time - last_pushed_ts) / 86400 )) # 86400 seconds in one day

        if [ "$age_in_days" -gt 30 ]; then
            delete_image "$org" "$repo" "$tag_name"
        else
            echo "${tag_name} is less than 30 days old. Not deleting."
        fi
    done
```

This suggestion ensures that images that are less than 30 days old still log a message to the console.

**Review by ajschmidt8 — APPROVED** [review]

**Pull request merged**